src/toxcore/onion/onion_announce.rs

```diff
@@ -0,0 +1,521 @@
+
+/// Number of seconds when onion ping id is valid after it was generated.
+/// To be precise ping id will be valid between `PING_ID_TIMEOUT` and
+/// 2 * `PING_ID_TIMEOUT` seconds.
+pub const PING_ID_TIMEOUT: u64 = 20;
+
+/// Number of seconds that announce entry can be stored in onion announce list
+/// without re-announcing.
+pub const ONION_ANNOUNCE_TIMEOUT: u64 = 300;
+
+/// Return number of seconds that have elapsed since Unix epoch.
+pub fn unix_time() -> u64 {
+    let start = SystemTime::now();
+    let since_the_epoch = start.duration_since(UNIX_EPOCH)
+        .expect("Current time is earlier than Unix epoch");
+    since_the_epoch.as_secs()
+}
+
+/** Entry that corresponds to announced onion node.
+
+When node successfully announce itself this entry is added to announced nodes
+list. It's considered expired after `ONION_ANNOUNCE_TIMEOUT` seconds.
+
+*/
+#[derive(Clone, Debug, Eq, PartialEq)]
+struct OnionAnnounceEntry {
+    /// Long term PublicKey of announced node
+    pub pk: PublicKey,
+    /// IP address of announced node
+    pub ret_ip_addr: IpAddr,
+    /// Port of announced node
+    pub ret_port: u16,
+    /// Onion return that should be used to send data packets to announced node
+    pub ret: OnionReturn,
+    /// PublicKey that should be used to encrypt data packets for announced node
+    pub data_pk: PublicKey,
+    /// Time when this entry was added to the list of announced nodes
+    pub time: u64
+}
+
+impl OnionAnnounceEntry {
+    /// Create new `OnionAnnounceEntry` object using current unix time.
+    pub fn new(pk: PublicKey, ret_ip_addr: IpAddr, ret_port: u16, ret: OnionReturn, data_pk: PublicKey) -> OnionAnnounceEntry {
+        OnionAnnounceEntry {
+            pk,
+            ret_ip_addr,
+            ret_port,
+            ret,
+            data_pk,
+            time: unix_time()
+        }
+    }
+
+    pub fn is_timed_out(&self) -> bool {
+        self.time + ONION_ANNOUNCE_TIMEOUT <= unix_time()
+    }
+}
+
+/// Size of serialized `OnionPingData` struct.
+const ONION_PING_DATA_SIZE: usize =
+    SECRET_BYTES_SIZE +
+    /* time */ 8 +
+    PUBLICKEYBYTES +
+    /* ip_type */ 1 +
+    /* ip_addr */ 8 +
+    /* port */ 2;
+
+/** Data on the basis of which onion ping id is calculated.
+
+Format of this struct is not specified by tox protocol and can be different in
+different implementations. That's possible because this struct is used for onion
+ping id generation and only node that generated it can verify it.
+
+Serialized form:
+
+Length   | Content
+-------- | ------
+`32`     | Secret bytes of onion node
+`8`      | Unix time in seconds divided by PING_ID_TIMEOUT
+`32`     | `PublicKey` of sender
+`1`      | IP type of sender
+`8`      | `IpAddr` of sender
+`2`      | Port of sender
+
+*/
+struct OnionPingData {
+    /// Secret bytes of onion node to make onion ping id unique
+    pub secret_bytes: [u8; SECRET_BYTES_SIZE],
+    /// Unix time in seconds
+    pub time: u64,
```

Comment:
`time` of?

Reply:
It's just a time, nothing specific. It's used for hash calculation and can be any time that user of this struct wants.

Reply:
OK. Put down your comment as doc